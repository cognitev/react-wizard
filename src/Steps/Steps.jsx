import React, { Component } from 'react';
import PropTypes from 'prop-types';
import equal from 'fast-deep-equal';
import WizardContext from '../Wizard/WizardContext';

/**
 * Wizard Steps wrapper
 */
class Steps extends Component {
  componentDidMount() {
    const { children: childs, context } = this.props;
    const { steps: wizardSteps, wizard } = context;

    const newSteps = React.Children.map(
      childs,
      ({ props: { children, render, ...config } }) => config,
    );

    if (wizardSteps && wizardSteps.length < 1 && childs.length > 0) {
      // while initializing component
      this.upliftSteps(wizard);
    } else {
      // TODO: check side Effects
      wizard.updateSteps(newSteps);
    }
  }

  shouldComponentUpdate() {
    const { context, children: childs } = this.props;
    const steps = React.Children.map(
      childs,
      ({ props: { children, render, ...config } }) => config,
    );
    /**
     * Do not do update in case steps were not equal
     */
    if (equal(steps, context.steps)) {
      return true;
    }
    return false;
  }

  componentDidUpdate() {
    const { context, children: childs } = this.props;
    const steps = React.Children.map(
      childs,
      ({ props: { children, render, ...config } }) => config,
    );
    if (!equal(steps, context.steps)) {
      // TODO: MAYBE issue of updating Wizard state while rendering Steps!!
      context.wizard.updateSteps(steps);
    }
  }

  /**
   * Update Wizard with Child Steps
   * @param wizard {object}
   */
  upliftSteps = (wizard) => {
    const { children: childs } = this.props;
    const steps = React.Children.map(
      childs,
      ({ props: { children, render, ...config } }) => config,
    );
    // TODO: issue of updating Wizard state while rendering Steps!!
    wizard.initializeWizard(steps);
  };

  getActiveStep = (context, children) => {
    // Render only activeStep from chil steps
    const { id: activeId } = context.step;
    const [activeChild = null] = React.Children.toArray(children).filter(
      ({ props: { id } }) => id === activeId,
    );
    return activeChild;
  }

  render() {
    const { children, context } = this.props;
    return this.getActiveStep(context, children);
  }
}

Steps.propTypes = {
  children: PropTypes.node,
  context: PropTypes.shape({
    wizard: PropTypes.shape({}),
    step: PropTypes.shape({}),
    steps: PropTypes.arrayOf(PropTypes.shape({})),
  }).isRequired,
};

Steps.defaultProps = {
  children: null,
};

/**
 * Export Steps with forwardedRef on the Steps Wrapper itself, and additional props from parent(s)
 */
export default React.forwardRef((props, ref) => (
  <WizardContext.Consumer>
    {({ steps, step, wizard }) => (
      <Steps
        {...props}
        context={{ steps, step, wizard }}
        ref={ref}
      >
        {props.children}
      </Steps>
    )}
  </WizardContext.Consumer>
));
