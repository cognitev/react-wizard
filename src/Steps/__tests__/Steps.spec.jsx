import React from 'react';
import { shallow } from 'enzyme';
import Steps from '../Steps';

// This is unit testing <Steps /> component, so no need for <Step /> functionality
const Step = () => null;

describe('<Steps /> Component', () => {
  it('should render correct child if controlled', () => {
    const rendered = shallow(
      <Steps>
        <Step id="step1" />
        <Step id="step2" />
        <Step id="step3" />
      </Steps>,
    );

    expect(rendered).toMatchSnapshot();
  });
});
