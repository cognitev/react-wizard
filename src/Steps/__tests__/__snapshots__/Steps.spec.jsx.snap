// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steps /> Component should render correct child if controlled 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ForwardRef>
    <Step
      id="step1"
    />
    <Step
      id="step2"
    />
    <Step
      id="step3"
    />
  </ForwardRef>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "step": Object {
            "enabled": null,
            "id": null,
            "onEnter": null,
            "onLeave": null,
          },
          "steps": Array [],
          "wizard": Object {},
        },
        "_currentValue2": Object {
          "step": Object {
            "enabled": null,
            "id": null,
            "onEnter": null,
            "onLeave": null,
          },
          "steps": Array [],
          "wizard": Object {},
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "step": Object {
              "enabled": null,
              "id": null,
              "onEnter": null,
              "onLeave": null,
            },
            "steps": Array [],
            "wizard": Object {},
          },
          "_currentValue2": Object {
            "step": Object {
              "enabled": null,
              "id": null,
              "onEnter": null,
              "onLeave": null,
            },
            "steps": Array [],
            "wizard": Object {},
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
